File: Java version/app/src/main/java/it/unical/mat/parsers/asp/ASPParser.java
package it.unical.mat.parsers.asp;

import it.unical.mat.parsers.asp.asp_parser_base.ASPGrammarBaseVisitor;
import it.unical.mat.parsers.asp.asp_parser_base.ASPGrammarLexer;
import it.unical.mat.parsers.asp.asp_parser_base.ASPGrammarParser;
import java.util.ArrayList;
import org.antlr.v4.runtime.BailErrorStrategy;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.ConsoleErrorListener;
import org.antlr.v4.runtime.DefaultErrorStrategy;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.atn.PredictionMode;

public class ASPParser extends ASPGrammarBaseVisitor <Void> {	
	private ArrayList <String> parameters = new ArrayList <> ();
	
	private ASPParser() {
		
    }
	
	public static final ASPParser parse(final String atom) {
		final CommonTokenStream tokens = new CommonTokenStream(new ASPGrammarLexer(CharStreams.fromString(atom)));
        final ASPGrammarParser parser = new ASPGrammarParser(tokens);
        final ASPParser visitor = new ASPParser();
        
        parser.getInterpreter().setPredictionMode(PredictionMode.SLL); 
        parser.removeErrorListeners();
        parser.setErrorHandler(new BailErrorStrategy());
     
        try {
        	visitor.visit(parser.output());
        } catch (final RuntimeException exception) {
        	if(exception.getClass() == RuntimeException.class && exception.getCause() instanceof RecognitionException) {
        		tokens.seek(0);
        		parser.addErrorListener(ConsoleErrorListener.INSTANCE);
        		parser.setErrorHandler(new DefaultErrorStrategy());
        		parser.getInterpreter().setPredictionMode(PredictionMode.LL); 
        		visitor.visit(parser.output());
        	}
        }
        
        return visitor;
	}
	
	public String [] getParameters() {
		return parameters.toArray(new String [0]);
	}
	
    @Override
    public Void visitTerm(ASPGrammarParser.TermContext ctx) {
    	parameters.add(ctx.getText());
    	
    	return null;
    }
}

File: CSharp version/it/unical/mat/parsers/asp/ASPParser.cs
using System;
using System.Collections.Generic;
using Antlr4.Runtime;
using Antlr4.Runtime.Atn;

namespace it.unical.mat.parsers.asp
{
    public class ASPParser : ASPGrammarBaseVisitor<object>
    {
        private readonly List<string> parameters = new List<string>();

        private ASPParser()
        {

        }

        public static ASPParser Parse(string atom)
        {
            CommonTokenStream tokens = new CommonTokenStream(new ASPGrammarLexer(CharStreams.fromstring(atom)));
            ASPGrammarParser parser = new ASPGrammarParser(tokens);
            ASPParser visitor = new ASPParser();
            parser.Interpreter.PredictionMode = PredictionMode.SLL;

            parser.RemoveErrorListeners();

            parser.ErrorHandler = new BailErrorStrategy();

            try
            {
                visitor.Visit(parser.output());
            }
            catch (SystemException exception)
            {
                if (exception.GetBaseException() is RecognitionException)
                {
                    tokens.Seek(0);
                    parser.AddErrorListener(ConsoleErrorListener<object>.Instance);

                    parser.ErrorHandler = new DefaultErrorStrategy();
                    parser.Interpreter.PredictionMode = PredictionMode.LL;

                    visitor.Visit(parser.output());
                }
            }

            return visitor;
        }

        public string[] GetParameters()
        {
            return parameters.ToArray();
        }

        public override object VisitTerm(ASPGrammarParser.TermContext context)
        {
            parameters.Add(context.GetText());

            return null;
        }
    }
}